### ISSUE TITLE
express 5.0.0

### ISSUE THREAD

**Issue by gireeshpunathil** [expressjs/discussions#100]
the original issue (PR https://github.com/expressjs/express/pull/2237) that tracks this is locked up for probable unmanageable number of discussions and some of which off-topic? I can't find an active issue that tracks this, so opening one.

I request to release express 5.

 - how was the user feedback on `5.0.0-alpha.7`? if it is reasonably positive, does it make sense to cut that into 5?
 - are there pending items in https://github.com/expressjs/express/pull/2237#issuecomment-570058548 that are real blockers? can they be pushed to in 5.1?
 - can we have a TC meeting to discuss this - challenges, gaps, actionables and owners?

I am not an express commiter, but a staunch fan. I can commit the necessary bandwidth required for upskilling myself and contribute to realize this!

**Comment by gireeshpunathil** [comment]
@dougwilson - 

> new path-to-regexp which will also need a corresponding section created in http://expressjs.com/en/guide/migrating-5.html

I would like to drive this to closure, any guidance on where to start or whom to contact?

> 4. Publish 5.0 beta

Why do we want a beta release? We ran into extra-ordinarily long iterations of alphas, and as there wasn't any concerns from its users, why can't we go straight upto 5.0.0 GA? anyways, we shall discuss this in the TC!

I am stating my opinion here that, releasing express 5.0.0 (as GA, not as beta) is the best thing to do, for the ecosystem.

> migration guide:

Who is / are the SMEs here? I would like to request that all the maintainers and triagers be `educated` on this topic - a couple of dedicated sessions if need be would be great. If you can share the info on  the SME part, I can co-ordinate with them on this space.

> how long do we want to support 4.x?

My take is that we defer it (taking a decision) to after the release of 5 and some insights from the field are gathered? Until then the standard 1 year applies, with a caveat of future amendments?

> Also we may want to comb over the issues in our deps

May be candidates for 6, 7, 8... ?

**Comment by dougwilson** [comment]
> I would like to drive this to closure, any guidance on where to start or whom to contact?

The path-to-regexp update has mainly been driven by @wesleytodd 

> Why do we want a beta release? We ran into extra-ordinarily long iterations of alphas, and as there wasn't any concerns from its users, why can't we go straight upto 5.0.0 GA?

We can absolutely do that, but it means that we drop adding anything new at this point and release alpha.7 as the GA. That is your suggestion, correct?

> Who is / are the SMEs here? I would like to request that all the maintainers and triagers be educated on this topic - a couple of dedicated sessions if need be would be great. If you can share the info on the SME part, I can co-ordinate with them on this space.

That is such a large question, I'm not even sure how to really answer it :) I mean, I would probably say in a general sense myself and @wesleytodd /shrug

> May be candidates for 6, 7, 8... ?

IMO one thing I really like about Express, and pretty much all my coworkers, is the _lack_ of constant breaking changes and majors. I would push back extremely strongly regarding this, unless you're talking about the far future. IMO I think we should not be releasing majors faster than Node.js itself, if even that. That means that if we release 5.0 today, then we're deferring everything not included into it until at least Feb 2021... If we want to do that, I think we should get a good idea on what exactly we are talking about here and the specifics around such a sudden change in direction that is so drastic.

**Comment by gireeshpunathil** [comment]
@dougwilson -

> We can absolutely do that, but it means that we drop adding anything new at this point and release alpha.7 as the GA. That is your suggestion, correct?

Yes.

> That means that if we release 5.0 today, then we're deferring everything not included into it until at least Feb 2021

Generally agree if there are immediate takers of the delta features - but as you rightly said, we need to look at those items to get a better feel. But here is a consumer's perspective: the more features get added to a major, the tougher it becomes to test, assess the impact, and migrate. I am not saying a 6 months cadence is good for this project, but on the other hand, too long gaps between majors make it hard to migrate too?

> I would push back extremely strongly regarding this

I guess this is where the real difference in opinion is, if at all - all others being just general sync up and ratification. How do we gather consensus on this? - **express release cadence**
 - deliberate in a separate issue
 - discuss in a dedicated meeting
 - gather user feedback

?

**Comment by gireeshpunathil** [comment]
I opened https://github.com/expressjs/express/issues/4197 for item 3 above

**Comment by dougwilson** [comment]
> I opened expressjs/express#4197 for item 3 above

I'm confused here: why ask a question for feedback on what to do, but then don't even wait for it? Why even ask the question "How do we gather consensus on this?" if you're not going to wait for any response?

**Comment by gireeshpunathil** [comment]
@dougwilson - I  acknowledge your confusion. I left item 1 and 2 to hear from you while initiated item 3, as it takes time to collect data; so when we discuss that, we have some data handy to look at?

**Comment by dougwilson** [comment]
Ok, it read to me on how do we do it a, b, or c? I think we need to first have the TC meeting so we can actually bring you to to speed on our past discussions on this topic so we can hear your and you hear ours and then, once there is a meeting of the minds, determine our path forward.

**Comment by gireeshpunathil** [comment]
ok.  I see there are 2 types of things:

 - items that needs action, no discussion / consensus required as such : (e.g:- path-to-regexp doc update) - can be worked out with community members, independently. 

 - items that needs discussion and consensus (e.g:- release cadence) - TC meeting to cover in its agenda.

**Comment by gireeshpunathil** [comment]
Just came up with a template for the release schedule, that we can have a look, brainstorm and bless in the next TC!

 - Feel free to add items that you believe needs to be met, by editing this comment.
 - If you are only proposing, probably fill only the first (and second?) columns
 - if you are both proposer and owner, fill all the columns
 - if you are owning the proposed work, fill the rest of the columns

**WEEK 1**

| item name | tracking issue | owner | challenge/support required | eta |
|-------------|----------------|-------|----------------------------|-----|
||||||
||||||

**WEEK 2**

| item name | tracking issue | owner | challenge/support required | eta |
|-------------|----------------|-------|----------------------------|-----|
||||||
||||||

**WEEK 3**

| item name | tracking issue | owner | challenge/support required | eta |
|-------------|----------------|-------|----------------------------|-----|
||||||
||||||

**WEEK 4**

| item name | tracking issue | owner | challenge/support required | eta |
|-------------|----------------|-------|----------------------------|-----|
|||||||
|||||||

**Comment by wesleytodd** [comment]
Can you elaborate on what `challenge/support required` means?

**Comment by gireeshpunathil** [comment]
@wesleytodd -

> Can you elaborate on what challenge/support required means?

`challenge`: anything that the `owner` wants to callout as a potential blocker in realizing the item
`support required`: anything that the `owner` request from the community members in help fulfilling the item

**Comment by wesleytodd** [comment]
Figured I would follow up here on the req/res stuff. It was on our short list which we I thought we all agreed to. The iteration for v6 is my biggest concern. If we think re importing at the existing files is not a breaking change I am good with that. I believe it should not break anything. The work for it is quick as I already have it half done. I can do it tonight if we are all in agreement. Are there any other concerns than just if it could be done in time?

**Comment by gireeshpunathil** [comment]
> If we think re importing at the existing files is not a breaking change I am good with that.

@wesleytodd - do you mean the approach @jonchurch suggested in https://github.com/expressjs/discussions/issues/106#issuecomment-595898871 ?

**Comment by gireeshpunathil** [comment]
re: https://github.com/expressjs/discussions/issues/100#issuecomment-596502248

given that it is a day's work, and it is not going to break anything, I am fine with that. thanks @wesleytodd !

**Comment by gireeshpunathil** [comment]
track the progress of items planned for the week:

 - extended query parser deprecation in 4.x lines: https://github.com/expressjs/express/pull/4208 raised, currently @dougwilson reviewing
 - 4.x (4.17.2 or 4.18?) release with the said deprecation in it. waiting on https://github.com/expressjs/express/pull/4208
 - repo captians proposal: https://github.com/expressjs/express/pull/4210
 - https://github.com/expressjs/express/issues/2432 : @wesleytodd going to PR soon
 - https://github.com/expressjs/express/pull/3621 - land as is (given the peer PR https://github.com/expressjs/express/pull/4208 in place), or re-PR?
 - @jonchurch  's 5 items : where do we stand?
 - https://github.com/pillarjs/router/pull/42: according to @dougwilson , it just needs re-branching and merging. @dougwilson  - are you planning to do that?
 - https://github.com/expressjs/expressjs.com/pull/1114 : needs review + landing. owners pls!

**Comment by wesleytodd** [comment]
Ok, digging back in after a while.  Here are the relevant links:

Original issue: https://github.com/expressjs/express/issues/2432
My first issue on the topic: https://github.com/expressjs/express/pull/3214
My follow up issue: https://github.com/expressjs/discussions/issues/82

Request repo: https://github.com/wesleytodd/express-request
Response repo: https://github.com/wesleytodd/express-response

Here are the open questions:

- Is there remaining objections to landing this in `5.0`?
  - My opinion is yes due to the reasons I made on the meeting: Makes http2/3 changes eaiser to land if it is decoupled from the main line express development between 5&6
- Should these be published as `@expressjs/request`/`@expressjs/response` or `@pillarjs/request`/`@pillarjs/response`?
  - I lean toward `@pillarjs` because these really would allow for other frameworks to have express compatible `req`/`res`.

From looking at the repos I have, this is the remaining todo list once we have answers to the above questions:

- Make sure the req and res in my repos are up-to-date with the 5.0 branch
- Move the tests fro res over and make sure they pass
- Move the repos into either `expressjs` or `pillarjs`
- Publish the modules
- Update PR #3214 to include the new references

**Comment by jonchurch** [comment]
@gireeshpunathil I've opened https://github.com/expressjs/express/pull/4212 just now. Might need a quick change or two pending feedback but it is moving along.

**Comment by gireeshpunathil** [comment]
@wesleytodd - re: https://github.com/expressjs/discussions/issues/100#issuecomment-596778171

> Is there remaining objections to landing this in 5.0? 

only @dougwilson had raised one concern on this, regarding breaking nature (which seem to have addressed now). @dougwilson - are you good with this now?

> Should these be published as @expressjs/request&@expressjs/response or @pillarjs/request&@pillarjs/response? 

I agree with pillarjs proposal, aiming to acheve cross-compatibility on this core module

**Comment by jonchurch** [comment]
@wesleytodd I'm happy to help with the req/res move. Let me know which of those issues you want to track this in. 

I can help move tests over or move documentation, whatever you want help with.

**Comment by wesleytodd** [comment]
> Move req/res out, but expose them as an export in the main express codebase that prints a deprecation warning if someone requires them

Sounds good to me!

> I agree with pillarjs proposal, aiming to acheve cross-compatibility on this core module
> pillarjs makes sense to me as well.

I will wait for @dougwilson's input for a bit, but if it is not opposition I will create the repos move my existing work into them. @jonchurch that would be a good place to collaborate, but if you have time before then and want to submit PRs against the repos linked above for whatever you have time on I will merge them before we move the repos.

**Comment by jonchurch** [comment]
@wesleytodd Assuming the move is green lit, transferring the repos is just a few clicks in github, if the repo owner (you) has permissions to create a repo in the destination.

So I think things can move forward on that front with the final placement happening whenever the decision gets made.

**Comment by gireeshpunathil** [comment]
I thougth we are going to have a TC sitting (my) today (12 hours from now) , but do not see a tracking issue. @jonchurch - are you going to setup one?

**Comment by dougwilson** [comment]
Yes @wesleytodd I think pillarjs org would make sense for them; since they are along the lines of "framework building blocks" and that is the goal of pillarjs, it makes the most sense to me. I could see it in expressjs as well, but so far that has been pretty much confined to express itself and middlewares. Along the lines of thinking of another framework using it, even for "Express compatibility" that also makes me think it should be in pillarjs.

It would be nice is there would be a place to discuss the actual architecture of those repos ideally. I think the biggest feedback I have from reviewing them is they... do not actually work independent of each other, yet they are published independently as if they could be. I'm not clear on if the issue is either they they depend on each other by accident or maybe that they should just be a single repo and not two separate repos.

The readme of `express-request` seems to imply it should be able to be used stand alone, but yet following the example and attempting to run it promptly fails due to there being no `this.app` defined. Trying to use `req.fresh` also fails, not just because `this.res` is not defined, but also because the `res` is expected to also have been wrapped with the separate package (which it itself has methods expecting `req` to be wrapped with the other package).

My point above is specifically _not_ to discuss issues / changes with those repos here, rather to give an example about something that would be nice to discuss somewhere, and putting it out for suggestion on what would be a place to have said discussions.

**Comment by gireeshpunathil** [comment]
3rd TC meeting happened on 18th https://github.com/expressjs/discussions/issues/109 with agenda as outlined in https://github.com/expressjs/discussions/issues/109#issuecomment-598560875

agreed action items:

 - @dougwilson to do the needful to merge all ready PRs (as listed in the above comment)
 - @wesleytodd to move forward with req/res refactoring: https://github.com/expressjs/discussions/issues/82
 - all to review @LinusU 's stream based Multer PR https://github.com/expressjs/multer/pull/399

**Comment by gireeshpunathil** [comment]
@jonchurch / @wesleytodd / @dougwilson 

could one of you pls spin up a TC meeting for this week?

**Comment by gireeshpunathil** [comment]
current status of ear-marked and priotized items:

https://github.com/pillarjs/router/pull/42 : landed

https://github.com/expressjs/express/pull/4208: ready to land since last 19 days
https://github.com/expressjs/express/pull/4210: ready to land since last 20 days
https://github.com/expressjs/express/pull/4212: under review

https://github.com/expressjs/expressjs.com/pull/1114: ready to land since last 20 days
https://github.com/pillarjs/request: repo created. not sure if anything else is pending

https://github.com/expressjs/expressjs.com/pull/1127: needs review, since 11 days
https://github.com/expressjs/expressjs.com/pull/1128: merged
https://github.com/expressjs/expressjs.com/pull/1129: needs review, since 4 days
https://github.com/expressjs/expressjs.com/pull/1130: needs review, since 4 days
https://github.com/expressjs/expressjs.com/pull/1131: needs review, since 10 days
https://github.com/expressjs/expressjs.com/pull/1132: needs review, since 9 days
https://github.com/expressjs/expressjs.com/pull/1134: needs review, since 4 days

can I request some focus on moving these please? (involves code reviews and merges) @expressjs/express-tc

**Comment by gireeshpunathil** [comment]
Reduced list after today's TC meeting:

https://github.com/expressjs/express/pull/3621: good to land
https://github.com/expressjs/express/pull/4208 : no concerns from anyone on not printing the deprecation. Close and abandon.
https://github.com/expressjs/express/pull/4212: needs review and progress
https://github.com/expressjs/expressjs.com/pull/1114: post release action
https://github.com/expressjs/expressjs.com/pull/1131: needs review and progress

@expressjs/express-tc - please chime in! we are getting closer, and this now needs few landings + the release process!

note 1: there isn't anything major between the last `alpha` and the current, so I strongly recommend for GA, not a beta.

note 2: the risk of delaying further is that new issues can slip in from unknown corners with potential of delaying and blocking things.

**Comment by gireeshpunathil** [comment]
https://github.com/expressjs/expressjs.com/pull/1131 is landed now, so we are down to 4!

**Comment by dougwilson** [comment]
In order to jump from alpha to GA, we will need to finish up the feature in the router repo, get the body parser 2.0 out to match the query praser change in 5.0, and get all those deps to be non beta release as well, as a GA express shouldn't have non GA deps. I believe the idea of a single beta is so we can release one intermediate express 5 between those, but if we want to hold the next 5 release for those things in order for it to be a GA release, we can certainly do so.

**Comment by gireeshpunathil** [comment]
@dougwilson - 

> In order to jump from alpha to GA, we will need to finish up the feature in the router repo, get the body parser 2.0 out to match the query praser change in 5.0, and get all those deps to be non beta release as well, as a GA express shouldn't have non GA deps.

this is surprising! where do this come from?  I have proposed GA 2 months back here https://github.com/expressjs/discussions/issues/100#issuecomment-590185515 and this was not raised!

> GA express shouldn't have non GA deps

where is this documented?

> we will need to finish up the feature in the router repo, get the body parser 2.0 out to match the query praser change in 5.0, and get all those deps to be non beta release as well, ...

are there issues / PRs opened for these? or need to start from scratch?

**Comment by dougwilson** [comment]
Yes, we know there is a lot of gaps in our docs, and I know myself and Wes have been trying hard to continue to push forward with the work needed and still try and provide as much information as possible. We are trying out best, but we do missing things. If you think it is desirable, we can go ahead and stop pushing forward on the 5.0 work and pivot into perhaps a summit-style type of thing to get everything that needs to be done across all three orgs completely documented.

Let me know what you think is a best use of time. We are here trying to best move everything forward, and with our limited resources we have at the moment. We built up our group quickly, with a sudden focus to release 5.0 as fast as possible, and perhaps we are cutting corners we should not, like fully articulating the plan.

I think perhaps we need to stop the work towards 5.0 at this point and focus on a full plan of all the sub-dependency work across the three orgs that we are holding in our minds so we can get a full picture of what is being done and the plan.

Does this sound like a good proposal? Do you have an alternative proposal?

As for the request of non-GA deps in GA code, we don't document it, but perhaps you're saying we should. Typically this is just, I believe, an expectation for Node.js modules; that when you `npm install express` it doesn't pull in a bunch of alpha- and beta-marked dependencies along with it when you didn't ask for an alpha- or beta-marked express? I feel like that would be quite surprising to see (GA code depending on non-GA code), but perhaps I am misunderstanding what it means to be GA? Perhaps we should fully document and articulate what going GA means before we proceed in order to understand that we meet the GA definition with our GA release 🤔

**Comment by gireeshpunathil** [comment]
@dougwilson - 

> We built up our group quickly, with a sudden focus to release 5.0 as fast as possible, and perhaps we are cutting corners we should not, like fully articulating the plan.

I wouldn't call it sudden - it has been 2 months since this issue is opened, to say the least. And there were several express 5.0 plans in the past as well?

> I think perhaps we need to stop the work towards 5.0 at this point and focus on a full plan of all the sub-dependency work across the three orgs that we are holding in our minds so we can get a full picture of what is being done and the plan.

No. Lot of contributors spent a lot of effort to get this out. Let us don't think of any other actions that will delay it any further. I am understanding now that there are several gaps in the documentation, process and other nagging problems that needs collaborative effort to improve upon, but let us don't allow those to derail the current focus and larger goal.

> we will need to finish up the feature in the router repo, get the body parser 2.0 out to match the query praser change in 5.0, and get all those deps to be non beta release as well, ...

are there issues / PRs opened for these? or need to start from scratch?

**Comment by dougwilson** [comment]
Yes, those repos have them as open issues and PR (mainly PRs targeting the milestone).

**Comment by gireeshpunathil** [comment]
https://github.com/pillarjs/router/pull/60 : PR to make `router 2.0` . All the checklists are marked, yet not sure what it is waiting for?

https://github.com/expressjs/body-parser/pull/66: PR to release `body-parser 2.0`: how this is a pre-requisite for `query praser change in 5.0`?

**Comment by dougwilson** [comment]
Hi @gireeshpunathil is it OK for me to delay the responses to your question until next week? I would like to ficus on 4.18 tonight/tomorrow if that is OK.

**Comment by gireeshpunathil** [comment]
I would like to resume this work. @dougwilson - when you get time, please let me know the latest list pre-requisites for 5!

**Comment by dougwilson** [comment]
Hi @gireeshpunathil the PR for the 5.0 release has a checklist tree that _should_ be up-to-date with that information. If not, please bring up what is outdated there. All the items we dropped from there in previous conversations have a strike out to indicate that they were removed from consideration on purpose to shrink the scope.

**Comment by gireeshpunathil** [comment]
I have extracted the outstanding items from https://github.com/expressjs/express/pull/2237#issue-18358589 as:

- [ ] Add support for Promises in app.param: https://github.com/pillarjs/router/pull/92
- [ ] Make query parser option default to 'simple' https://github.com/expressjs/express/pull/3621
- [ ] Make res.status throw on invalid argument type https://github.com/expressjs/express/pull/4212
- [ ] New path matching syntax https://github.com/expressjs/express/pull/4321
- [ ] Make bodyParser.urlencoded default to 'simple': https://github.com/expressjs/body-parser/pull/402

with the last one opened today, seeing no other open PRs on it. Next I will review each of the existing PRs to see where do they stand, and what actions required to move those forward

**Comment by dougwilson** [comment]
I am locking this issue until the communications issues between myself and @gireeshpunathil are resolved.

**Comment by UlisesGascon** [comment]
Next phase is in https://github.com/expressjs/discussions/issues/233
